---
title: "ALL-10X"
output: html_document
date: "2023-02-10"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Seurat

```{r}
library("Seurat")
```


```{r}
seurat_object <- Read10X(
  "/data/MS_2022_scRNASeq_CLIP/cellranger/swALL/outs/per_sample_outs/swALL/count/sample_feature_bc_matrix/",
  gene.column = 2,
  cell.column = 1,
  unique.features = TRUE,
  strip.suffix = T
)
```

```{r}
#head(seurat_object)
```
```{r}
#head(seurat_object$`Antibody Capture`)
```
Making sure we have the same cells in the same order for both assays.
```{r}
all.equal(colnames(seurat_object$`Gene Expression`), colnames(seurat_object$`Antibody Capture`))
```
# CreateSeuratObject and integrate ADT data
```{r}
ALL_seurat_object <- CreateSeuratObject(counts = seurat_object$`Gene Expression`)
```

```{r}
# create a new assay to store ADT information
adt_assay <- CreateAssayObject(counts = seurat_object$`Antibody Capture`)

# add this assay to the previously created Seurat object
ALL_seurat_object[["ADT"]] <- adt_assay

# Validate that the object now contains multiple assays
Assays(ALL_seurat_object)
```

```{r}
rownames(ALL_seurat_object[["ADT"]])
```

```{r}
DefaultAssay(ALL_seurat_object) <- "RNA"
```

```{r}
ALL_seurat_object[["percent.mt"]] <- PercentageFeatureSet(ALL_seurat_object, pattern = "^MT-")
```

```{r}
# Visualize QC metrics as a violin plot
VlnPlot(ALL_seurat_object, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```
```{r}
ALL_seurat_object <- subset(ALL_seurat_object, subset = nFeature_RNA > 200 & nFeature_RNA < 4000 & percent.mt < 20)
```

4714 -> 4405
```{r}
# Visualize QC metrics as a violin plot
VlnPlot(ALL_seurat_object, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```


```{r message=FALSE, warning=FALSE}
ALL_seurat_object <- NormalizeData(ALL_seurat_object)
ALL_seurat_object <- FindVariableFeatures(ALL_seurat_object)
ALL_seurat_object <- ScaleData(ALL_seurat_object)
ALL_seurat_object <- RunPCA(ALL_seurat_object, verbose = FALSE)
ALL_seurat_object <- FindNeighbors(ALL_seurat_object, dims = 1:30)
ALL_seurat_object <- FindClusters(ALL_seurat_object, resolution = 0.8, verbose = FALSE)
ALL_seurat_object <- RunUMAP(ALL_seurat_object, dims = 1:30)

```
```{r}
DimPlot(ALL_seurat_object, label = TRUE)
```
# Creating a flowFrame object to be do gates based on CD14-CD19 protein expression

```{r, eval= F}
library(flowCore)
# Create a new flowFrame object
flow_data <- new("flowFrame", exprs = as.matrix(ALL_seurat_object@assays$ADT@data), description = list(rownames(ALL_seurat_object)))
# Write the flowFrame object to an FCS file
write.FCS(flow_data, file = "/data/workspace/new.fcs")
```

## Filter BY GATES

```{r}
gates_filtered_indices <- readRDS("/data/MS_2023_scRNASeq_CLIP_ALL_George/indices_filtered.rds")
```
Select only the gated cells
```{r}
gates_filtered_seurat <- ALL_seurat_object[,gates_filtered_indices]
```
Export the barcodes to be used later for dynamo
```{r}
write.csv(colnames(gates_filtered_seurat),"/data/MS_2023_scRNASeq_CLIP_ALL_George/gated_events_barcodes.csv")
```
Write the gene expression matrix
```{r}
write.csv(gates_filtered_seurat@assays$RNA@scale.data, "/data/MS_2023_scRNASeq_CLIP_ALL_George/gene_expression_matrix.csv")
```
Save the seurat PCA embedding based on gene expression matrix
```{r, eval= F}
pca_embedding_after_seurat <-gates_filtered_seurat@reductions$pca@cell.embeddings 
saveRDS(pca_embedding_after_seurat,"/data/MS_2023_scRNASeq_CLIP_ALL_George/pca_embed_after_seurat.RDS")
```
###
```{r}
library("SingleR")
library(celldex)
```

```{r}
ref.data <- DatabaseImmuneCellExpressionData()
```

#Matrix 1.5.3
```{r}
library(Matrix)
predictions <- SingleR(test=as.SingleCellExperiment(gates_filtered_seurat), 
    ref=ref.data, labels=ref.data$label.main)
```

```{r}
gates_filtered_seurat[["SingleR.labels"]] <- as.factor(predictions$labels)
```

```{r}
DimPlot(gates_filtered_seurat, group.by = "SingleR.labels", pt.size = 0.3)
```
```{r, eval= F}
saveRDS(as.factor(predictions$labels), "/data/MS_2023_scRNASeq_CLIP_ALL_George/singleR_annotation.RDS")
```
# Velocyto
```
velocyto run -m /media/data1/Reference/Cellranger/hg38_rmsk.gtf sample_alignments.bam /media/data1/Reference/Cellranger/refdata-gex-GRCh38-2020-A/genes/genes.gtf 
-b /media/data1/MS_2023_scRNASeq_CLIP_ALL_George/barcodes.tsv

```
# Dynamo

Refer to:
/media/data1/MS_2023_scRNASeq_CLIP_ALL_George/velocyto/10x_SWALL_analysis.ipynb

Output cells should be re-ordered the same order as Seurat (alphabetically).
# Using the gates for labels in tviblindi

## Modify original functions

```{r, eval= F}
library(tviblindi)
```

```{r, eval= F}


fcstable<-
  function (fcs, table_file = "channels.txt") 
  {
    fcs <- read.FCS(fcs,emptyValue = FALSE)
    res <- fcs@parameters@data
    res <- data.frame(res, use = 0)
    write.table(res, file = table_file, col.names = TRUE, row.names = FALSE, 
                sep = "\t")
    return(invisible(res))
  }

tviblindi_from_flowjo1<-
  function (fcsn, wsp, fcstable, eventsufix = "_G[0-9]+$", eventprefix = "/[A-Z]_", 
            origin = "^A_", normalize = FALSE, sep = "\t", ftrans = function(x) asinh(x/5), 
            comp = NULL, sampNloc = "sampleNode") 
  {
    require(CytoML)
    require(flowWorkspace)
    ws <- open_flowjo_xml(wsp)
    protilatky <- read.table(fcstable, header = TRUE, sep = sep)
    fcs <- flowCore::read.FCS(fcsn,emptyValue = FALSE)
    
    if (!is.null(comp)) 
      fcs <- compensate(fcs, fcs@description[[comp]])
    chans <- as.character(protilatky$name[protilatky$use == 1])
    COO_colnames <- as.character(protilatky$desc[protilatky$use == 1])
    print("hu1")
    gs <- flowjo_to_gatingset(ws, name = 1, subset = 1, sampNloc = sampNloc,emptyValue=FALSE)
    print("hu")
    PP <- gs_get_pop_paths(gs[[1]])
    ppi <- grep(eventsufix, PP)
    if (length(ppi) == 0) {
      cat("Suffix not found - returning NULL", "\n")
      return(NULL)
    }
    cat("events used: ", PP[ppi], "\n")
    events_sel <- NULL
    for (i in ppi) events_sel <- c(events_sel, which(gh_pop_get_indices(gs[[1]], 
                                                                        PP[i[1]])))
    events_sel <- unique(events_sel)
    efcs <- flowCore::exprs(fcs)
    colnames(efcs) <- as.character(colnames(efcs))
    COO <- efcs[events_sel, chans]
    colnames(COO) <- COO_colnames
    COO <- ftrans(COO)
    if (!is.null(normalize)) {
      if (normalize == "perc") 
        COO <- normalize.perc(COO)
      else if (normalize == "scale") 
        COO <- scale(COO)
    }
    gate_ev_non <- list()[1:length(eventprefix)]
    j <- 0
    for (pref in eventprefix) {
      print(pref)
      j <- j + 1
      leafs <- grep(pref, PP)
      indM <- NULL
      for (ii in leafs) indM <- cbind(indM, as.numeric(gh_pop_get_indices_mat(gs[[1]], 
                                                                              PP[ii])))
      colnames(indM) <- gsub(".*/", "", PP[leafs])
      gate_ev <- colnames(indM)[apply(indM, MARGIN = 1, which.max)]
      ss <- which(rowSums(indM) == 0)
      gate_ev[ss] <- "ungated"
      gate_ev_non[[j]] <- gate_ev[events_sel]
    }
    names(gate_ev_non) <- paste("labels", 1:length(eventprefix))
    tv1 <- tviblindi(COO, labels = gate_ev_non, fcs_path = fcsn, 
                     events_sel = events_sel)
    if (!is.null(origin)) {
      stems <- levels(tv1$labels[[1]])[grep(origin, levels(tv1$labels[[1]]))]
      Set_origin(tv1, stems)
      if (is.null(tv1$origin)) 
        stop("Wrong population of origin")
    }
    return(tv1)
  }


```

# Load gating labels

```{r, eval=FALSE}

fcstable("/data/fcs_ADT/filtered/filtered_adt.fcs",table_file = "/data/fcs_ADT/filtered/channels.txt")
tv0<-tviblindi_from_flowjo1(wsp = "/data/fcs_ADT/filtered/swALL_10x_gates_filtered.wsp",
                            fcsn = "/data/fcs_ADT/filtered/filtered_adt.fcs",eventsufix = "swALL$",
                            fcstable = "/data/fcs_ADT/filtered/channels.txt",origin = NULL,eventprefix = "[A-C]_.*")
#tv0$labels[[1]]
```

```{r, eval= F}
data<-read.csv("/data/10X_analysis/X_pca.csv") #PCA embedding of Dynamo
group_id<-tv0$labels[[1]]
tv3<-tviblindi(data=as.matrix(data),labels=group_id)
DimRed(tv3)
DimRed(tv3,method="umap")
Set_origin(tv3,label = "A_B_blasts",origin_name = "A_B_blasts")
KNN(tv3)
Som(tv3,xdim = 15,ydim = 15) #kmeans clustering by default - 15*15 clusters
Filtration(tv3) #default setting is too conservative, less simplices could be created with same resolution (e.g. Filtration(tv3,alpha2=1))
Pseudotime(tv3,weighted = TRUE,origin_name = "A_B_blasts") 
Walks(tv3,N=1000,origin_name = "A_B_blasts") 
#tv3$vae$save("/data/10X_analysis/model.json", "/data/10X_analysis/model.h5")
##Adding un-rooted RNA velocity method
velocity <- read.csv("/data/10X_analysis/Velocity_pca.csv")
pseudotime_no_root <- get_pseudotime_from_velocity(tv3, 30, MatrixOfVelocity=as.matrix(velocity))
tv3$pseudotime$calculatedPseudotimeNoRoot$res<-as.numeric(pseudotime_no_root)
tv3$origin$calculatedPseudotimeNoRoot<-which.min(tv3$pseudotime$calculatedPseudotimeNoRoot$res)
Walks(tv3,N=1000,origin_name = "calculatedPseudotimeNoRoot")


singler_labels <- readRDS("10X_analysis/singleR_annotation.RDS")
tv3$labels<-singler_labels
launch_shiny(tv3)

```

